File: test/algorithms.jl
# Test the accuracy of twofloat's main algorithms
# Relative error bounds: cf. paper from J.M. Muller et al.
# Function inputs/outputs dataset is created by /test/algorithms.ts

using Test
using JSON

setprecision(BigFloat, 512)

const TwoF64 = Tuple{Float64, Float64}

struct ArgsList
    op1::Vector{Tuple{Float64}}
    op2::Vector{Tuple{TwoF64}}
    op11::Vector{Tuple{Float64, Float64}}
    op21::Vector{Tuple{TwoF64, Float64}}
    op22::Vector{Tuple{TwoF64, TwoF64}}
end

struct TestSet
    argsList::ArgsList
    fnOutput::Dict{String, Vector{TwoF64}}
end

json = read("$(@__DIR__)/algorithms-testset.json", String)
testset = JSON.parse(json, TestSet; null=NaN)

args_list = testset.argsList
fn_output = testset.fnOutput

coverage = Dict(keys(fn_output) .=> false)

const split_max = prevfloat(floatmax(Float64)/(2^27 + 1))
const u = big(2.0)^-precision(Float64)
const ε₀ = big(eps(0.0))

println()
@testset verbose = true "Error-Free Transforms ───" begin ######################

    @testset "split" begin
        output = fn_output["split"]
        args = args_list.op1
        coverage["split"] = true
        @test length(args) == length(output)
        for (i, (x,)) in enumerate(args)
            zhi, zlo = output[i]
            if abs(x) > split_max
                # overflow
                @test isnan(zhi + zlo)
            else
                @test x == zhi + zlo == big(zhi) + big(zlo)
            end
        end
    end

    @testset "normalize" begin
        output = fn_output["normalize"]
        args = args_list.op11
        coverage["normalize"] = true
        @test length(args) == length(output)
        for (i, (x, y)) in enumerate(args)
            zhi, zlo = output[i]
            z = big(zhi) + big(zlo)
            r = big(x) + big(y)
            if abs(z) > floatmax(Float64)
                # overflow
                @test !isfinite(zhi + zlo)
            else
                @test z == r
            end
        end
    end

    @testset "twoSum" begin
        output = fn_output["twoSum"]
        args = args_list.op11
        coverage["twoSum"] = true
        @test length(args) == length(output)
        for (i, (x, y)) in enumerate(args)
            zhi, zlo = output[i]
            z = big(zhi) + big(zlo)
            r = big(x) + big(y)
            if abs(z) > floatmax(Float64)
                # overflow
                @test !isfinite(zhi + zlo)
            else
                @test z == r
            end
        end
    end

    @testset "twoProd" begin
        output = fn_output["twoProd"]
        args = args_list.op11
        coverage["twoProd"] = true
        @test length(args) == length(output)
        for (i, (x, y)) in enumerate(args)
            zhi, zlo = output[i]
            z = big(zhi) + big(zlo)
            r = big(x) * big(y)
            if abs(r) > floatmax(Float64) || abs(x) > split_max || abs(y) > split_max
                # overflow
                @test !isfinite(zhi + zlo)
            elseif abs(zlo) < floatmin(Float64)
                # most likely underflow
                @test abs(z - r) < 1.5ε₀
            else
                @test z == r
            end
        end
    end
end

println()
@testset verbose = true "Error bounds op21 ───────" begin ######################
    args = args_list.op21

    @testset "DWPlusFP" begin
        output = fn_output["DWPlusFP"]
        coverage["DWPlusFP"] = true
        @test length(args) == length(output)
        rel_err = 2u^2
        abs_err = r -> max(abs(rel_err * r), ε₀)
        for (i, ((xhi, xlo), y)) in enumerate(args)
            zhi, zlo = output[i]
            z = big(zhi) + big(zlo)
            r = big(xhi) + big(y) + big(xlo)
            if abs(r) > floatmax(Float64)
                # overflow
                @test !isfinite(zhi + zlo)
            else
                @test abs(z - r) < abs_err(r)
            end
        end
    end

    @testset "DWTimesFP1" begin
        output = fn_output["DWTimesFP1"]
        coverage["DWTimesFP1"] = true
        @test length(args) == length(output)
        rel_err = 1.5u^2 + 4u^3
        abs_err = r -> max(abs(rel_err * r), 2ε₀)
        for (i, ((xhi, xlo), y)) in enumerate(args)
            zhi, zlo = output[i]
            z = big(zhi) + big(zlo)
            r = big(xhi)*big(y) + big(xlo)*big(y)
            if abs(r) > floatmax(Float64)
                # overflow
                @test !isfinite(zhi + zlo)
            else
                @test abs(z - r) < abs_err(r)
            end
        end
    end

    @testset "DWDivFP3" begin
        output = fn_output["DWDivFP3"]
        coverage["DWDivFP3"] = true
        @test length(args) == length(output)
        rel_err = 3u^2
        abs_err = r -> max(abs(rel_err * r), 2ε₀)
        for (i, ((xhi, xlo), y)) in enumerate(args)
            zhi, zlo = output[i]
            z = big(zhi) + big(zlo)
            r = big(xhi)/big(y) + big(xlo)/big(y)
            if abs(r) > floatmax(Float64)
                # overflow
                @test !isfinite(zhi + zlo)
            else
                @test abs(z - r) < abs_err(r)
            end
        end
    end
end

println()
@testset verbose = true "Error bounds op22 ───────" begin ######################
    args = args_list.op22

    @testset "AccurateDWPlusDW" begin
        output = fn_output["AccurateDWPlusDW"]
        coverage["AccurateDWPlusDW"] = true
        @test length(args) == length(output)
        rel_err = 3u^2 + 13u^3
        abs_err = r -> max(abs(rel_err * r), ε₀)
        for (i, ((xhi, xlo), (yhi, ylo))) in enumerate(args)
            zhi, zlo = output[i]
            z = big(zhi) + big(zlo)
            r = big(xhi) + big(yhi) + big(xlo) + big(ylo)
            if abs(r) > floatmax(Float64)
                # overflow
                @test !isfinite(zhi + zlo)
            else
                @test abs(z - r) < abs_err(r)
            end
        end
    end

    @testset "DWTimesDW1" begin
        output = fn_output["DWTimesDW1"]
        coverage["DWTimesDW1"] = true
        @test length(args) == length(output)
        rel_err = 5u^2
        abs_err = r -> max(abs(rel_err * r), 2.5ε₀)
        for (i, ((xhi, xlo), (yhi, ylo))) in enumerate(args)
            zhi, zlo = output[i]
            z = big(zhi) + big(zlo)
            r = (big(xhi) + big(xlo)) * (big(yhi) + big(ylo))
            if abs(r) > floatmax(Float64)
                # overflow
                @test !isfinite(zhi + zlo)
            else
                @test abs(z - r) < abs_err(r)
            end
        end
    end

    @testset "DWDivDW2" begin
        output = fn_output["DWDivDW2"]
        coverage["DWDivDW2"] = true
        @test length(args) == length(output)
        rel_err = 15u^2 + 56u^3
        abs_err = r -> max(abs(rel_err * r), 2.5ε₀)
        for (i, ((xhi, xlo), (yhi, ylo))) in enumerate(args)
            zhi, zlo = output[i]
            z = big(zhi) + big(zlo)
            r = (big(xhi) + big(xlo)) / (big(yhi) + big(ylo))
            if abs(r) > floatmax(Float64)
                # overflow
                @test !isfinite(zhi + zlo)
            else
                @test abs(z - r) < abs_err(r)
            end
        end
    end
end

println()
@testset verbose = true "Error bounds op1 ────────" begin ######################
    args = args_list.op1

    @testset "twoInv" begin
        output = fn_output["twoInv"]
        coverage["twoInv"] = true
        @test length(args) == length(output)
        rel_err = 3u^2
        abs_err = r -> max(abs(rel_err * r), ε₀)
        for (i, (x,)) in enumerate(args)
            zhi, zlo = output[i]
            z = big(zhi) + big(zlo)
            r = inv(big(x))
            if abs(r) > floatmax(Float64)
                # overflow
                @test !isfinite(zhi + zlo)
            else
                @test abs(z - r) < abs_err(r)
            end
        end
    end
end

println()
@testset verbose = true "Error bounds op11 ───────" begin ######################
    args = args_list.op11

    @testset "twoDiv" begin
        output = fn_output["twoDiv"]
        coverage["twoDiv"] = true
        @test length(args) == length(output)
        rel_err = 3u^2
        abs_err = r -> max(abs(rel_err * r), ε₀)
        for (i, (x, y)) in enumerate(args)
            zhi, zlo = output[i]
            z = big(zhi) + big(zlo)
            r = big(x) / big(y)
            if abs(r) > floatmax(Float64)
                # overflow
                @test !isfinite(zhi + zlo)
            else
                @test abs(z - r) < abs_err(r)
            end
        end
    end
end

println()
@testset verbose = true "Error bounds op2 ────────" begin ######################
    args = args_list.op2

    @testset "DWInv" begin
        output = fn_output["DWInv"]
        coverage["DWInv"] = true
        @test length(args) == length(output)
        rel_err = 15u^2 + 56u^3
        abs_err = r -> max(abs(rel_err * r), ε₀)
        for (i, ((xhi, xlo),)) in enumerate(args)
            zhi, zlo = output[i]
            z = big(zhi) + big(zlo)
            r = big(1.0) / (big(xhi) + big(xlo))
            if abs(r) > floatmax(Float64)
                # overflow
                @test !isfinite(zhi + zlo)
            else
                @test abs(z - r) < abs_err(r)
            end
        end
    end
end

###

println()
@testset "Algorithm coverage ──────" begin
    for (fn, covered) in coverage
        @test (fn, covered) == (fn, true)
    end
end

File: test/algorithms.ts
/**
 * Pre-test for the main algorithms
 *
 * Generates inputs and collect outputs for each algorithm then export the whole
 * as JSON so we can import it and do proper tests in Julia using BigFloat.
 */

import {
  f64,
  TwoF64,
  normalize as _normalize,
  split,
  twoSum,
  twoProd,
} from '../src/index';

import { exponent } from '@lvlte/ulp';
import fs from 'node:fs';

import {
  DWPlusFP,
  AccurateDWPlusDW,
  DWTimesFP1,
  DWTimesDW1,
  DWDivFP3,
  DWDivDW2,
} from '../src/base/algorithms';

type Expand<T> = {} & { [P in keyof T]: Expand<T[P]> };

type _UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (
  (k: infer I) => void
) ? I : never;

type UnionToIntersection<U> = _UnionToIntersection<U> extends infer O
  ? { [K in keyof O]: O[K] }
  : never;

interface FnSig {
  'op1': (x: f64) => TwoF64;
  'op11': (x: f64, y: f64) => TwoF64;
  'op21': (x: TwoF64, y: f64) => TwoF64;
  'op22': (x: TwoF64, y: TwoF64) => TwoF64;
}

// Wrap normalize so it is tested with the |x| ≥ |y| condition satisfied
const normalize: typeof _normalize = (x, y) => {
  return Math.abs(x) >= Math.abs(y) ? _normalize(x, y) : _normalize(y, x);
};

// Functions to test grouped by signature
const fnBySig = {
  'op1': {split},
  'op11': {normalize, twoSum, twoProd},
  'op21': {DWPlusFP, DWTimesFP1, DWDivFP3},
  'op22': {AccurateDWPlusDW, DWTimesDW1, DWDivDW2},
} satisfies {
  [K in keyof FnSig]: { [fnName: string]: FnSig[K] }
};

type FnBySig = typeof fnBySig;
type TestedFunctions = UnionToIntersection<FnBySig[keyof FnBySig]>;
type FnName = keyof TestedFunctions;

type ArgsListBySig = { [K in keyof FnBySig]: Parameters<FnSig[K]>[] };
type FnOutputList = { [K in FnName]: ReturnType<TestedFunctions[K]>[] }

// Pseudo-random number generator
const SEED = Math.sqrt(2);
const random = (function () {
  let n = SEED;
  return function(): number {
    return Math.abs(Math.sin(n++));
  }
})();

// Lists of arguments (grouped by FnSig) to pass to the TestedFunctions
const argsList: ArgsListBySig = {'op1': [], 'op11': [], 'op21': [], 'op22': []};

// Exponent range for the generated numbers
const [emin, emax] = [-128, 127];

// Fill argsList with number combinations in the domain [±2^emin, ±2^emax]
for (let e1 = emin; e1 <= emax; e1+=3) {
  for (let e2 = emin+1; e2 <= emax; e2+=3) {
    for (const [s1, s2] of [[1, 1], [1, -1], [-1, 1], [-1, -1]]) {
      const sw = random();
      const sx = random();
      const sy = random();
      const sz = random();

      const w = s1 * sw * 2**(e1 - exponent(sw));
      const x = s1 * sx * 2**(e1 - exponent(sx));
      const y = s2 * sy * 2**(e2 - exponent(sy));
      const z = s2 * sz * 2**(e2 - exponent(sz));

      const xy = normalize(x, y);
      const wz = normalize(w, z);

      argsList['op1'].push([z]);
      argsList['op11'].push([x, y]);
      argsList['op21'].push([xy, z]);
      argsList['op22'].push([xy, wz]);
    }
  }
}

// Produce the list of outputs keyed by function given argsList
const fnOutput = {} as FnOutputList;
for (const sid in fnBySig) {
  const fnGroup = fnBySig[sid as keyof FnBySig];
  const argsGroup = argsList[sid as keyof FnBySig]
  for (const fnName in fnGroup) {
    fnOutput[fnName as FnName] = [];
    const fn = (fnGroup as TestedFunctions)[fnName as FnName];
    const fnOut = fnOutput[fnName as FnName]!;
    for (const args of argsGroup) {
      // @ts-ignore (TS doesn't understand correlated unions)
      const result = fn(...args);
      fnOut.push(result);
    }
  }
}

// Inputs/Outputs object
const testset = { argsList, fnOutput} as Expand<{
  argsList: ArgsListBySig, fnOutput: FnOutputList
}>;

// Export as JSON
const testsetJSON = JSON.stringify(testset);
fs.writeFileSync('test/algorithms-testset.json', testsetJSON, 'utf8');
